File: src/main/webapp/resources/js/service/incidenceService.js
/*
 * Copyright (C) 2015 Pablo Grela Palleiro (pablogp_9@hotmail.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
'use strict';

angular.module('membersApp').factory('IncidenceService', [ '$http', '$q', function($http, $q) {

	var REST_SERVICE_URI = '/members/incidenceList/';
	var csrf = '?' + document.getElementById("csrf.parameterName").value + '=' + document.getElementById("csrf.token").value;

	var factory = {
		fetchAllIncidences : fetchAllIncidences,
		fetchAllIncidencesClose : fetchAllIncidencesClose,
		incidenceUp : incidenceUp,
		incidenceDown : incidenceDown,
		incidenceAnswer : incidenceAnswer,
	};

	return factory;

	function fetchAllIncidences() {
		var deferred = $q.defer();
		$http.get(REST_SERVICE_URI + csrf).then(function(response) {
			deferred.resolve(response.data);
		}, function(errResponse) {
			console.error('Error while fetching Incidences');
			deferred.reject(errResponse);
		});
		return deferred.promise;
	}

	function fetchAllIncidencesClose() {
		var deferred = $q.defer();
		$http.get(REST_SERVICE_URI + "close/" + csrf).then(function(response) {
			deferred.resolve(response.data);
		}, function(errResponse) {
			console.error('Error while fetching Incidences Close');
			deferred.reject(errResponse);
		});
		return deferred.promise;
	}
	
	function incidenceUp(id) {
		var deferred = $q.defer();
		var url = REST_SERVICE_URI + 'incidenceUp/' + id + csrf;
		$http.post(url).then(function(response) {
			deferred.resolve(response.data);
		}, function(errResponse) {
			console.error('Error while up incidence');
			deferred.reject(errResponse);
		});
		return deferred.promise;
	}

	function incidenceDown(id) {
		var deferred = $q.defer();
		var url = REST_SERVICE_URI + 'incidenceDown/' + id + csrf;
		$http.post(url).then(function(response) {
			deferred.resolve(response.data);
		}, function(errResponse) {
			console.error('Error while down incidence');
			deferred.reject(errResponse);
		});
		return deferred.promise;
	}

	function incidenceAnswer(id, answer) {
		var deferred = $q.defer();
		answer =  "&answer=" + answer;
		var url = REST_SERVICE_URI + 'incidenceAnswer/' + id  + csrf + answer;
		$http.post(url).then(function(response) {
			deferred.resolve(response.data);
		}, function(errResponse) {
			console.error('Error while answer incidence');
			deferred.reject(errResponse);
		});
		return deferred.promise;
	}

} ]);

File: src/main/java/org/cuacfm/members/model/incidenceservice/IncidenceService.java
/**
 * Copyright (C) 2015 Pablo Grela Palleiro (pablogp_9@hotmail.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.cuacfm.members.model.incidenceservice;

import java.util.List;

import org.cuacfm.members.model.account.Account;
import org.cuacfm.members.model.exceptions.ExistPaymentsException;
import org.cuacfm.members.model.exceptions.UniqueException;
import org.cuacfm.members.model.incidence.Incidence;
import org.cuacfm.members.model.incidence.IncidenceDTO;
import org.springframework.web.multipart.MultipartFile;

/** The Class IncidenceService. */
public interface IncidenceService {

	/**
	 * Save an training into database.
	 *
	 * @param incidence the training
	 * @return Incidence
	 * @throws UniqueException the unique exception
	 */
	public Incidence save(Incidence incidence, MultipartFile[] files);

	/**
	 * Update Incidence.
	 *
	 * @param incidence the incidence
	 * @return Incidence
	 * @throws UniqueException the unique exception
	 */
	public Incidence update(Incidence incidence);

	/**
	 * Delete.
	 *
	 * @param incidence the incidence
	 * @throws ExistPaymentsException the exist payments exception
	 */
	public void delete(Incidence incidence);

	/**
	 * Delete.
	 *
	 * @param incidence the incidence
	 * @param account the account
	 * @throws ExistPaymentsException the exist payments exception
	 */
	public void delete(Incidence incidence, Account account) throws ExistPaymentsException;

	/**
	 * Find by Name the incidence.
	 *
	 * @param name the name
	 * @return Incidence
	 */
	public Incidence findByName(String name);

	/**
	 * Find by id returns incidence which has this identifier.
	 *
	 * @param id the id
	 * @return incidence
	 */
	public Incidence findById(Long id);

	/**
	 * Get all incidences.
	 *
	 * @return List<Incidence>
	 */
	public List<Incidence> getIncidenceList();

	/**
	 * Get all active incidences.
	 *
	 * @return List<Incidence>
	 */
	public List<Incidence> getIncidenceListActive();

	/**
	 * Gets the incidence list by user.
	 *
	 * @param account the account
	 * @return the incidence list by user
	 */
	public List<Incidence> getIncidenceListByUser(Account account);

	/**
	 * Gets the incidence list close.
	 *
	 * @return the incidence list close
	 */
	public List<Incidence> getIncidenceListClose();

	/**
	 * Up.
	 *
	 * @param incidence the incidence
	 */
	public void up(Incidence incidence);

	/**
	 * Down.
	 *
	 * @param incidence the incidence
	 */
	public void down(Incidence incidence);

	/**
	 * Answer.
	 *
	 * @param incidence the incidence
	 * @param answer the answer
	 */
	public void answer(Incidence incidence, String answer);

	/**
	 * Gets the incidences DTO.
	 *
	 * @param incidences the incidences
	 * @return the incidences DTO
	 */
	public List<IncidenceDTO> getIncidencesDTO(List<Incidence> incidences);

	/**
	 * Gets the incidence DTO.
	 *
	 * @param incidence the incidence
	 * @return the incidence DTO
	 */
	public IncidenceDTO getIncidenceDTO(Incidence incidence);
}
